README
DBIx::DBSchema

Copyright (c) 2000-2007 Ivan Kohler
Copyright (c) 2000 Mail Abuse Prevention System LLC
Copyright (c) 2007-2013 Freeside Internet Services, Inc.
All rights reserved.
This program is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

This module implements an OO-interface to database schemas.  Using this module,
you can create a database schema with an OO Perl interface.  You can read the
schema from an existing database.  You can save the schema to disk and restore
it from different process.  You can write SQL CREATE statements statements for
different databases from a single source.  You can transform one schema to
another, adding any necessary new columns, tables, indices and foreign keys.

Currently supported databases are MySQL, PostgreSQL, and SQLite.  Sybase and
Oracle drivers are partially implemented.  DBIx::DBSchema will attempt to use
generic SQL syntax for other databases.  Assistance adding support for other
databases is welcomed.  See the DBIx::DBSchema::DBD manpage, "Driver Writer's
Guide and Base Class".

To install:
	perl Makefile.PL
	make
	make test # nothing substantial yet
	make install

Documentation will then be available via `man DBIx::DBSchema' or
`perldoc DBIx::DBSchema'.

Anonymous CVS access is available:
  $ export CVSROOT=":pserver:anonymous@cvs.420.am:/home/cvs/cvsroot"
  $ cvs login
  (Logging in to anonymous@cvs.420.am)
  CVS password: anonymous
  $ cvs checkout DBIx-DBSchema
as well as <http://www.420.am/cgi-bin/viewvc.cgi/DBIx-DBSchema>.

Homepage: <http://www.420.am/dbix-dbschema>



Changes
Revision history for Perl module DBIx::DBSchema

0.44 2013-11-15 17:54:37 PST
    - POD fixes from Xavier Guimard <x.guimard@free.fr> and Damyan Ivanov
      <dmn@debian.org> of the Debian Perl Group, closes: CPAN#82187
    - More foreign key support
      + Table alteration (removing foreign keys)
      + Fix MATCH / ON DELETE / ON UPDATE

0.43 2013-11-04 14:58;32 PST
    - Incorporate CPAN::Changes::Spec-compliant Changes file from Neil Bowers
      <neilb@cpan.org>, closes: CPAN#90023, thanks!
    - Fix table creation broken by 0.42
    - Create/alter tables before foreign key changes

0.42 2013-11-03 01:09:27 PST
    - Basic foreign key support
      + table creation
      + table alteration (adding new foreign keys)
      + reverse-engineering (Pg driver)

0.41_01 not released
    - consolidate multiple ADD/ALTER COLUMN statements into one ALTER TABLE

0.40 2011-12-17 17:03:51 PST
    - doc: sql_update_schema link to sql_add_column misspelled
    - Pg: fix mapping for DOUBLE PRECISION type to avoid needless alter cols
    - mysql: implement altering column types & lengths

0.39 2010-03-26 20:24:58 PDT
    - mysql: TEXT->LONGTEXT
    - mysql: when reverse engineering, transform a default of
      CURRENT_TIMESTAMP into the more common NOW()
    - mysql: fix reverse-engineering of empty string default

0.38 2010-01-14 15:26:13 PST
    - Bump version number for ->quoted_default availability & default
      reverse-engineering changes

0.37 not released
    - Patch from Slavin Rezic <srezic@cpan.org> to prevent quoting around
      numeric defaults in Pg.
    - Pg: use default_db_schema when adding SERIAL columns instead of
      hardcoding "public".
    - Pg: Initial support for handling changes to a column's type or size.
    - Case sensitivity fix for Slavin's patch to prevent quoting around
      numeric defaults in Pg.
    - Column default values: refactor handling, improve Pg reverse
      engineering and implement schema changes.

0.36 2007-12-13 17:49:35 PST
    - Patch from ISHIGAKI@cpan.org to suppress unnecessary warnings about
      undefined local_options, thanks!
    - fix bug preventing adding new columns as primary keys under Pg

0.35 2007-10-29 18:58:36 PDT
    - Fix minor breakage (pretty_print) resulting from Jesse's changes.
    - Update mysql driver to handle BIGSERIAL columns
    - Update Column.pm, move all mysql and Pg-specific code to DBD driver
      callbacks
    - Update Table.pm, add local_options
    - Fix mysql NULL reverse-engineering and updating code

0.34 2007-08-19 10:08:51 PDT
    - More work on update schema from Slaven Rezic <srezic@cpan.org>,
      thanks!
      + implement table dropping (closes: CPAN#27936)
      + implement column dropping (closes: CPAN#27896)
    - Fix to quiet warnings from internal use of old API from Jesse Vincent
      <jesse+cpan@fsck.com>, thanks! (closes: CPAN#27958)
    - Make table dropping optional, not the default.

0.33 2007-06-28 18:46:15 PDT
    - Overhaul of index representation: indices (both normal and unique)
      now have names and are DBIx::DBSchema::Index objects
    - update_schema now handles indices!
    - Bump version numbers in Table.pm, Column.pm and DBD.pm
    - Pg reverse-engineering fix for column order in multi-column indices,
      to prevent needless drop/add of identical indices
    - mysql reverse-engineering patch from Brian Phillips
      <bphillips@cpan.org>, closes: CPAN#17582, thanks!
    - mysql NAME vs NAME_lc patch from Ralf Hack <ralf@beetlecraft.net>,
      closes: CPAN#16715, thanks!
    - mysql fix for additional column data from Chris Mungall
      <cjm@fruitfly.org>, closes: CPAN#20859, thanks!
    - SQLite SERIAL patch from IN SUK JOUNG <i.joung@gmail.com>, and fix
      for mis-application of said patch from Slaven Rezic
      <srezic@cpan.org>, thanks!
    - Update README wrt current CVS info and URL, closes: CPAN#27577

0.32 2007-04-18 15:02:25 PDT
    - increment the version numbers in Column.pm and Table.pm and the
      "use" statements accordingly
    - Error reporting for load constructor
    - Update documentation wrt supported databases and new update_schema
      stuff
    - Fixes for dropping nullability on old Pg (<= 7.2)
    - Fixes for adding nullability on old Pg (<= 7.3)
    - Throw a warning if pg_server_version is unavailable, assume >= 7.3
    - fix POD error in DBSchema::DBD::SQLite that confused pod2man and
      prevented the documentation from being installed.  Patch from Niko
      Tyni <ntyni@iki.fi>, thanks!

0.31 2006-03-30 05:28:20 PST
    - more schema update stuff:
    - added Column::sql_alter_column
    - added Table::sql_alter_table
    - added DBSchema::sql_update_schema and DBSchema::update_schema

0.30 2006-02-16 16:43:01 PST
    - "Too much uptime"
    - Remove buggy debugging from Column.pm
    - Remove removed TODO from MANIFEST

0.29 2006-02-16 13:54:42 PST
    - Column::sql_add_column fix when adding primary keys to Pg 7.2.x
    - workaround for PAUSE parsing of DBIx::DBSchema::DBD::Pg version: move
      DBD::Pg verison checking after $VERSION declaration, thanks Andreas!
    - kludge: allow scalar ref default to force quoting off, to add things
      like functions and empty values as defaults
    - Move TODO file to DBSchema.pm and DBS/Column.pm BUGS sections

0.28 2005-11-30 09:46:47 PST
    - Initial SQLite support from Jesse Vincent
    - fix typo in DBIx::DBSchema::DBD POD doc

0.27 2005-08-15 23:31:54 PDT
    - MySQL patch for enum types from Andy Orr
    - new Column::sql_add_column method!

0.26 2005-04-07 01:09:53 PDT
    - ask for "public" db schema only from Pg

0.25 2005-04-06 16:12:38 PDT
    - depend on DBD::Pg 1.32 or 1.41+ (1.40 was bunk)

0.24 2005-03-11 02:20:55 PST
    - Oracle driver from Daniel Hanks <hanksdc@about-inc.com> and Peter
      Bowen <pbowen@aboutws.com>.
    - Switch from FreezeThaw to Storable, keep ability to read old files

0.23 2004-02-16 17:35:54 PST
    - Update Pg dependancy to 1.32

DBSchema/DBD/Pg.pm
package DBIx::DBSchema::DBD::Pg;
use base qw(DBIx::DBSchema::DBD);

use strict;
use DBD::Pg 1.32;

our $VERSION = '0.19';

die "DBD::Pg version 1.32 or 1.41 (or later) required--".
    "this is only version $DBD::Pg::VERSION\n"
  if $DBD::Pg::VERSION != 1.32 && $DBD::Pg::VERSION < 1.41;

our %typemap = (
  'BLOB'           => 'BYTEA',
  'LONG VARBINARY' => 'BYTEA',
  'TIMESTAMP'      => 'TIMESTAMP WITH TIME ZONE',
);

=head1 NAME

DBIx::DBSchema::DBD::Pg - PostgreSQL native driver for DBIx::DBSchema

=head1 SYNOPSIS

use DBI;
use DBIx::DBSchema;

$dbh = DBI->connect('dbi:Pg:dbname=database', 'user', 'pass');
$schema = new_native DBIx::DBSchema $dbh;

=head1 DESCRIPTION

This module implements a PostgreSQL-native driver for DBIx::DBSchema.

=cut

sub default_db_schema  { 'public'; }

sub columns {
  my($proto, $dbh, $table) = @_;
  my $sth = $dbh->prepare(<<END) or die $dbh->errstr;
    SELECT a.attname, t.typname, a.attlen, a.atttypmod, a.attnotnull,
           a.atthasdef, a.attnum
    FROM pg_class c, pg_attribute a, pg_type t
    WHERE c.relname = '$table'
      AND a.attnum > 0 AND a.attrelid = c.oid AND a.atttypid = t.oid
    ORDER BY a.attnum
END
  $sth->execute or die $sth->errstr;

  map {

    my $type = $_->{'typname'};
    $type = 'char' if $type eq 'bpchar';

    my $len = '';
    if ( $_->{attlen} == -1 && $_->{atttypmod} != -1 
         && $_->{typname} ne 'text'                  ) {
      $len = $_->{atttypmod} - 4;
      if ( $_->{typname} eq 'numeric' ) {
        $len = ($len >> 16). ','. ($len & 0xffff);
      }
    }

    my $default = '';
    if ( $_->{atthasdef} ) {
      my $attnum = $_->{attnum};
      my $d_sth = $dbh->prepare(<<END) or die $dbh->errstr;
        SELECT substring(d.adsrc for 128) FROM pg_attrdef d, pg_class c
        WHERE c.relname = '$table' AND c.oid = d.adrelid AND d.adnum = $attnum
END
      $d_sth->execute or die $d_sth->errstr;

      $default = $d_sth->fetchrow_arrayref->[0];

      if ( _type_needs_quoting($type) ) {
        $default =~ s/::([\w ]+)$//; #save typecast info?
        if ( $default =~ /^'(.*)'$/ ) {
          $default = $1;
          $default = \"''" if $default eq '';
        } else {
          my $value = $default;
          $default = \$value;
        }
      } elsif ( $default =~ /^[a-z]/i ) { #sloppy, but it'll do
        my $value = $default;
        $default = \$value;
      }

    }

    [
      $_->{'attname'},
      $type,
      ! $_->{'attnotnull'},
      $len,
      $default,
      ''  #local
    ];

  } @{ $sth->fetchall_arrayref({}) };
}

sub primary_key {
  my($proto, $dbh, $table) = @_;
  my $sth = $dbh->prepare(<<END) or die $dbh->errstr;
    SELECT a.attname, a.attnum
    FROM pg_class c, pg_attribute a, pg_type t
    WHERE c.relname = '${table}_pkey'
      AND a.attnum > 0 AND a.attrelid = c.oid AND a.atttypid = t.oid
END
  $sth->execute or die $sth->errstr;
  my $row = $sth->fetchrow_hashref or return '';
  $row->{'attname'};
}

sub unique {
  my($proto, $dbh, $table) = @_;
  my $gratuitous = { map { $_ => [ $proto->_index_fields($dbh, $_ ) ] }
      grep { $proto->_is_unique($dbh, $_ ) }
        $proto->_all_indices($dbh, $table)
  };
}

sub index {
  my($proto, $dbh, $table) = @_;
  my $gratuitous = { map { $_ => [ $proto->_index_fields($dbh, $_ ) ] }
      grep { ! $proto->_is_unique($dbh, $_ ) }
        $proto->_all_indices($dbh, $table)
  };
}

sub _all_indices {
  my($proto, $dbh, $table) = @_;
  my $sth = $dbh->prepare(<<END) or die $dbh->errstr;
    SELECT c2.relname
    FROM pg_class c, pg_class c2, pg_index i
    WHERE c.relname = '$table' AND c.oid = i.indrelid AND i.indexrelid = c2.oid
END
  $sth->execute or die $sth->errstr;
  map { $_->{'relname'} }
    grep { $_->{'relname'} !~ /_pkey$/ }
      @{ $sth->fetchall_arrayref({}) };
}

sub _index_fields {
  my($proto, $dbh, $index) = @_;
  my $sth = $dbh->prepare(<<END) or die $dbh->errstr;
    SELECT a.attname, a.attnum
    FROM pg_class c, pg_attribute a, pg_type t
    WHERE c.relname = '$index'
      AND a.attnum > 0 AND a.attrelid = c.oid AND a.atttypid = t.oid
    ORDER BY a.attnum
END
  $sth->execute or die $sth->errstr;
  map { $_->{'attname'} } @{ $sth->fetchall_arrayref({}) };
}

sub _is_unique {
  my($proto, $dbh, $index) = @_;
  my $sth = $dbh->prepare(<<END) or die $dbh->errstr;
    SELECT i.indisunique
    FROM pg_index i, pg_class c, pg_am a
    WHERE i.indexrelid = c.oid AND c.relname = '$index' AND c.relam = a.oid
END
  $sth->execute or die $sth->errstr;
  my $row = $sth->fetchrow_hashref or die 'guru meditation #420';
  $row->{'indisunique'};
}

#using this
#******** QUERY **********
#SELECT conname,
#  pg_catalog.pg_get_constraintdef(r.oid, true) as condef
#FROM pg_catalog.pg_constraint r
#WHERE r.conrelid = '16457' AND r.contype = 'f' ORDER BY 1;
#**************************

# what's this do?
#********* QUERY **********
#SELECT conname, conrelid::pg_catalog.regclass,
#  pg_catalog.pg_get_constraintdef(c.oid, true) as condef
#FROM pg_catalog.pg_constraint c
#WHERE c.confrelid = '16457' AND c.contype = 'f' ORDER BY 1;
#**************************

sub constraints {
  my($proto, $dbh, $table) = @_;
  my $sth = $dbh->prepare(<<END) or die $dbh->errstr;
    SELECT conname, pg_catalog.pg_get_constraintdef(r.oid, true) as condef
      FROM pg_catalog.pg_constraint r
        WHERE r.conrelid = ( SELECT oid FROM pg_class
                               WHERE relname = '$table'
                                 AND pg_catalog.pg_table_is_visible(oid)
                           )
          AND r.contype = 'f'
END
  $sth->execute;

  map { $_->{condef}
          =~ /^FOREIGN KEY \(([\w\, ]+)\) REFERENCES (\w+)\(([\w\, ]+)\)\s*(.*)$/
            or die "unparsable constraint: ". $_->{condef};
        my($columns, $table, $references, $etc ) = ($1, $2, $3, $4);
        +{ 'constraint' => $_->{conname},
           'columns'    => [ split(/,\s*/, $columns) ],
           'table'      => $table,
           'references' => [ split(/,\s*/, $references) ],
           #XXX $etc not handled yet for MATCH, ON DELETE, ON UPDATE
         };
      }
    grep $_->{condef} =~ /^\s*FOREIGN\s+KEY/,
      @{ $sth->fetchall_arrayref( {} ) };
}

sub add_column_callback {
  my( $proto, $dbh, $table, $column_obj ) = @_;
  my $name = $column_obj->name;

  my $pg_server_version = $dbh->{'pg_server_version'};
  my $warning = '';
  unless ( $pg_server_version =~ /\d/ ) {
    $warning = "WARNING: no pg_server_version!  Assuming >= 7.3\n";
    $pg_server_version = 70300;
  }

  my $hashref = { 'sql_after' => [], };

  if ( $column_obj->type =~ /^(\w*)SERIAL$/i ) {

    $hashref->{'effective_type'} = uc($1).'INT';

    #needs more work for old Pg?
      
    my $nextval;
    warn $warning if $warning;
    if ( $pg_server_version >= 70300 ) {
      my $db_schema  = default_db_schema();
      $nextval = "nextval('$db_schema.${table}_${name}_seq'::text)";
    } else {
      $nextval = "nextval('${table}_${name}_seq'::text)";
    }

    push @{ $hashref->{'sql_after'} }, 
      "ALTER TABLE $table ALTER COLUMN $name SET DEFAULT $nextval",
      "CREATE SEQUENCE ${table}_${name}_seq",
      "UPDATE $table SET $name = $nextval WHERE $name IS NULL",
    ;

  }

  if ( ! $column_obj->null ) {
    $hashref->{'effective_null'} = 'NULL';

    warn $warning if $warning;
    if ( $pg_server_version >= 70300 ) {

      push @{ $hashref->{'sql_after'} },
        "ALTER TABLE $table ALTER $name SET NOT NULL";

    } else {

      push @{ $hashref->{'sql_after'} },
        "UPDATE pg_attribute SET attnotnull = TRUE ".
        " WHERE attname = '$name' ".
        " AND attrelid = ( SELECT oid FROM pg_class WHERE relname = '$table' )";

    }

  }

  $hashref;

}

sub alter_column_callback {
  my( $proto, $dbh, $table, $old_column, $new_column ) = @_;
  my $name = $old_column->name;

  my %canonical = (
    'SMALLINT'         => 'INT2',
    'INT'              => 'INT4',
    'BIGINT'           => 'INT8',
    'SERIAL'           => 'INT4',
    'BIGSERIAL'        => 'INT8',
    'DECIMAL'          => 'NUMERIC',
    'REAL'             => 'FLOAT4',
    'DOUBLE PRECISION' => 'FLOAT8',
    'BLOB'             => 'BYTEA',
    'TIMESTAMP'        => 'TIMESTAMPTZ',
  );
  foreach ($old_column, $new_column) {
    $_->type($canonical{uc($_->type)}) if $canonical{uc($_->type)};
  }

  my $pg_server_version = $dbh->{'pg_server_version'};
  my $warning = '';
  unless ( $pg_server_version =~ /\d/ ) {
    $warning = "WARNING: no pg_server_version!  Assuming >= 7.3\n";
    $pg_server_version = 70300;
  }

  my $hashref = {};

  #change type
  if ( ( $canonical{uc($old_column->type)} || uc($old_column->type) )
         ne ( $canonical{uc($new_column->type)} || uc($new_column->type) )
       || $old_column->length ne $new_column->length
     )
  {

    warn $warning if $warning;
    if ( $pg_server_version >= 80000 ) {

      $hashref->{'sql_alter_type'} =
        "ALTER COLUMN ". $new_column->name.
        " TYPE ". $new_column->type.
        ( ( defined($new_column->length) && $new_column->length )
              ? '('.$new_column->length.')'
              : ''
        )

    } else {
      warn "WARNING: can't yet change column types for Pg < version 8\n";
    }

  }

  # change nullability from NOT NULL to NULL
  if ( ! $old_column->null && $new_column->null ) {

    warn $warning if $warning;
    if ( $pg_server_version < 70300 ) {
      $hashref->{'sql_alter_null'} =
        "UPDATE pg_attribute SET attnotnull = FALSE
          WHERE attname = '$name'
            AND attrelid = ( SELECT oid FROM pg_class
                               WHERE relname = '$table'
                           )";
    }

  }

  # change nullability from NULL to NOT NULL...
  # this one could be more complicated, need to set a DEFAULT value and update
  # the table first...
  if ( $old_column->null && ! $new_column->null ) {

    warn $warning if $warning;
    if ( $pg_server_version < 70300 ) {
      $hashref->{'sql_alter_null'} =
        "UPDATE pg_attribute SET attnotnull = TRUE
           WHERE attname = '$name'
             AND attrelid = ( SELECT oid FROM pg_class
                                WHERE relname = '$table'
                            )";
    }

  }

  $hashref;

}

sub column_value_needs_quoting {
  my($proto, $col) = @_;
  _type_needs_quoting($col->type);
}

sub _type_needs_quoting {
  my $type = shift;
  $type !~ m{^(
               int(?:2|4|8)?
             | smallint
             | integer
             | bigint
             | (?:numeric|decimal)(?:\(\d+(?:\s*\,\s*\d+\))?)?
             | real
             | double\s+precision
             | float(?:\(\d+\))?
             | serial(?:4|8)?
             | bigserial
             )$}ix;
}


=head1 AUTHOR

Ivan Kohler <ivan-dbix-dbschema@420.am>

=head1 COPYRIGHT

Copyright (c) 2000 Ivan Kohler
Copyright (c) 2000 Mail Abuse Prevention System LLC
Copyright (c) 2009-2013 Freeside Internet Services, Inc.
All rights reserved.
This program is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=head1 BUGS

columns doesn't return column default information.

=head1 SEE ALSO

L<DBIx::DBSchema>, L<DBIx::DBSchema::DBD>, L<DBI>, L<DBI::DBD>

=cut 

1;

